### ISSUE TITLE
Parser does not support NVARCHAR2

### ISSUE THREAD

**Issue by kooshball** [jOOQ/jOOQ#9065]
I am getting the following error when trying to run the code generation using the "rg.jooq.meta.extensions.ddl.DDLDatabase":

```
C:\FAST\JDK64\1.8.0.212\bin\java.exe -Dmaven.multiModuleProjectDirectory=C:\ProgramData\IdeaProjects\CommitmentApp\cmtt-oracle "-Dmaven.home=C:\FAST\Apache Maven\3.6.0" "-Dclassworlds.conf=C:\FAST\Apache Maven\3.6.0\bin\m2.conf" -javaagent:C:\FAST\IntelliJUltimate\2019.1.3\lib\idea_rt.jar=59298:C:\FAST\IntelliJUltimate\2019.1.3\bin -Dfile.encoding=UTF-8 -classpath "C:\FAST\Apache Maven\3.6.0\boot\plexus-classworlds-2.5.2.jar" org.codehaus.classworlds.Launcher -Didea.version2019.1.3 generate-sources -P jooq
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------< com.jpmchase.ccb.hlt.corr.commitment:cmtt-oracle >----------
[INFO] Building cmtt-oracle 1.0.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- jooq-codegen-maven:3.11.12:generate (generate-jooq-sources) @ cmtt-oracle ---
[INFO] No <inputCatalog/> was provided. Generating ALL available catalogs instead.
[INFO] License parameters       
[INFO] ----------------------------------------------------------
[INFO]   Thank you for using jOOQ and jOOQ's code generator
[INFO]                          
[INFO] Database parameters      
[INFO] ----------------------------------------------------------
[INFO] Reading from: src\main\resources\db\oracle-schema.sql [*]
[ERROR] An exception occurred while parsing script source : src/main/resources/db/oracle-schema.sql. Please report this error to https://github.com/jOOQ/jOOQ/issues/new
org.jooq.impl.ParserException: Unknown data type: [7:10] ...rnal generator.
CREATE TABLE ACCNTING 
(
  DBCODE [*]NVARCHAR2(8) NOT NULL
, ENTTYPE NVARCHAR2(20) NOT NULL 
, GPRICE NUMBER(18, 3) 
...
    at org.jooq.impl.ParserContext.exception (ParserImpl.java:8513)
    at org.jooq.impl.ParserImpl.parseDataTypePrefix (ParserImpl.java:7262)
    at org.jooq.impl.ParserImpl.parseDataType (ParserImpl.java:7023)
    at org.jooq.impl.ParserImpl.parseCreateTable (ParserImpl.java:2296)
    at org.jooq.impl.ParserImpl.parseCreate (ParserImpl.java:1834)
    at org.jooq.impl.ParserImpl.parseQuery (ParserImpl.java:674)
    at org.jooq.impl.ParserImpl.parse (ParserImpl.java:445)
    at org.jooq.impl.ParserImpl.parse (ParserImpl.java:432)
    at org.jooq.meta.extensions.ddl.DDLDatabase.load (DDLDatabase.java:207)
    at org.jooq.meta.extensions.ddl.DDLDatabase.load (DDLDatabase.java:181)
    at org.jooq.meta.extensions.ddl.DDLDatabase.create0 (DDLDatabase.java:141)
    at org.jooq.meta.AbstractDatabase.create (AbstractDatabase.java:231)
    at org.jooq.meta.AbstractDatabase.create (AbstractDatabase.java:220)
    at org.jooq.meta.AbstractDatabase.getDialect (AbstractDatabase.java:202)
    at org.jooq.codegen.AbstractGenerator.logDatabaseParameters (AbstractGenerator.java:155)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:299)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:752)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:222)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:197)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  15.522 s
[INFO] Finished at: 2019-08-20T09:07:11-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jooq.trial:jooq-codegen-maven:3.11.12:generate (generate-jooq-sources) on project cmtt-oracle: Error running jOOQ code generation tool: Unknown data type: [7:10] ...rnal generator.
[ERROR] CREATE TABLE ACCNTING 
[ERROR] (
[ERROR]   DBCODE [*]NVARCHAR2(8) NOT NULL
[ERROR] , ENTTYPE NVARCHAR2(20) NOT NULL 
[ERROR] , GPRICE NUMBER(18, 3) 
[ERROR] ...
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

Process finished with exit code 1
```

The create script is:

```sql
CREATE TABLE ACCNTING 
(
  DBCODE NVARCHAR2(8) NOT NULL
, ENTTYPE NVARCHAR2(20) NOT NULL 
, GPRICE NUMBER(18, 3) 
, NPRICE NUMBER(18, 3) 
, GSADJ NUMBER(18, 3) 
, GMADJ NUMBER(18, 3) 
, NSADJ NUMBER(16, 6) 
, NMADJ NUMBER(18, 3) 
, WFEES NUMBER(18, 3) 
, INCNTVE NCHAR(1) 
, TBLFNDG NCHAR(1) 
, WTRNSFR NCHAR(1) 
, ORDRMIC NCHAR(1) 
, CLOSENME NCHAR(1) 
, WAREHSE NCHAR(1) 
, CLSLOAN NCHAR(1) 
, TAXID NVARCHAR2(20) 
, INCORPFLG NCHAR(1) 
, INCORPST NCHAR(2) 
, ARMLINE NUMBER(18, 3) 
, FIXDLINE NUMBER(16, 6) 
, SVCPORT NUMBER(18, 3) 
, SMFCPORT NUMBER(18, 3) 
, SVCDATE DATE 
, SMFCDATE DATE 
, NETWORTH NUMBER(18, 3) 
, NWDATE DATE 
, WLSTATUS NVARCHAR2(8) 
, INTLIMIT NUMBER(16, 6) 
, PUBLIMIT NUMBER(16, 6) 
, SELLCL NVARCHAR2(8) 
, SCLSDATE DATE 
, CRAGENCY NVARCHAR2(25) 
, PERNETWORTH NUMBER(18, 3) 
, PERNWDATE DATE 
, S_CLOSTYPE NVARCHAR2(8) 
, WLDATE DATE 
, USEPARENT NCHAR(1) 
, PAYEECODE NVARCHAR2(30) 
, PAYCODEID NVARCHAR2(3) 
, LAST_UPDATE_USER NVARCHAR2(30) 
, LAST_UPDATED DATE 
) 
```

My pom has:

```xml
    <profiles>
        <profile>
            <id>jooq</id>
            <build>
                <plugins>
                    <plugin>
                        <groupId>org.jooq.trial</groupId>
                        <artifactId>jooq-codegen-maven</artifactId>
                        <version>${jooq.version}</version>
                        <executions>
                            <execution>
                                <id>generate-jooq-sources</id>
                                <phase>generate-sources</phase>
                                <goals>
                                    <goal>generate</goal>
                                </goals>
                                <configuration>
                                    <generator>
                                        <generate>
                                            <deprecated>false</deprecated>
                                            <daos>true</daos>
                                            <interfaces>true</interfaces>
                                            <springAnnotations>true</springAnnotations>
                                            <daos>true</daos>
                                            <pojosEqualsAndHashCode>true</pojosEqualsAndHashCode>
                                            <javaTimeTypes>true</javaTimeTypes>
                                        </generate>
                                        <database>
                                            <name>org.jooq.meta.extensions.ddl.DDLDatabase</name>
                                            <inputCatalog></inputCatalog>
                                            <inputSchema>GFSMORT</inputSchema>
                                            <outputSchemaToDefault>true</outputSchemaToDefault>
                                            <outputCatalogToDefault>true</outputCatalogToDefault>
                                            <properties>
                                                <property>
                                                    <key>sort</key>
                                                    <value>semantic</value>
                                                </property>
                                                <property>
                                                    <key>scripts</key>
                                                    <value>src/main/resources/db/oracle-schema.sql</value>
                                                </property>
                                            </properties>
                                        </database>
                                        <target>
                                            <clean>true</clean>
                                            <packageName>com.jpmchase.ccb.hlt.corr.commitment.model.jooq.oracle</packageName>
                                            <directory>target/generated-sources</directory>
                                        </target>
                                    </generator>
                                </configuration>
                            </execution>
                        </executions>
                        <dependencies>
                            <dependency>
                                <groupId>org.jooq.trial</groupId>
                                <artifactId>jooq-meta-extensions</artifactId>
                                <version>${jooq.version}</version>
                            </dependency>
                        </dependencies>
                    </plugin>
                </plugins>
            </build>
        </profile>
    </profiles>
```

When I switch NVARCHAR2 to VARCHAR2, I don't get any errors.  I see in the API documentation that NVARCHAR is acceptable.  Shouldn't NVARCHAR2 be also valid?

**Comment by lukaseder** [comment]
Thank you very much for your report. Our parser seems to currently not support `NVARCHAR2` as a data type.

A possible workaround is to replace the data type by `VARCHAR2` or by `NVARCHAR` prior to passing it to the `DDLDatabase`, e.g. by using the `maven-replacer-plugin`.

We'll fix this right away.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12. Backport scheduled for 3.11.13 (#9068)

Thanks again for reporting!
